### ISSUE TITLE
Long running PHP script

### ISSUE THREAD

**Issue by clanstyles** [deuill/go-php#56]
How would you handle a long running PHP script? How would we get "sigterm" notifications to allow the PHP script to close gracefully?

**Comment by clanstyles** [comment]
Hey @deuill thanks for the response!

I like the use of a class deconstructor to control the shutting down, I hadn't thought of that. In additional tests, it doesn't seem like you can create multiple instances of a script, even if you share the context without a segfault / memory corruption. Do you have any examples of that?

**Comment by deuill** [comment]
Correct, running concurrent scripts in the same context is treacherous, even if the two share no global scope (class names, function names, global/file-level variables etc).

There is no real way to avoid this without using PHP with ZTS, which isn't currently supported here (I "temporarily" removed ZTS support during the move to PHP7, as it was significantly refactored between PHP5 and PHP7). I'd be glad to help if you or anyone else wanted to go down the rabbit-hole of integrating ZTS back in, however most of the things I've used `go-php` for are limited to one execution thread (as stupid and limiting that may seem).

**Comment by clanstyles** [comment]
Yeah I'd be interested in helping. I have a use case where I would prefer to keep mpst of the code in Go but run a library in PHP. I need to be able to run many instances isolated. 

Where do we start? Whats required to get ZTS support working?

**Comment by deuill** [comment]
While the way PHP handles thread-safety has been cleaned up between PHP 5.x and 7.x (starting with [this commit](https://github.com/php/php-src/commit/bdeb220f48825642f84cdbf3ff23a30613c92e86)), the principle is the same: configuring PHP with `--enable-maintainer-zts` will define the `ZTS` macro, which is checked during compilation.

Documentation on PHP internals is miserable, and the code isn't documented for the most part either, so the way I've mostly developed through this library is trying to find equivalent functionality in the PHP source code.

Take, for example, the [`engine_init` function](https://github.com/deuill/go-php/blob/dbea71e0cf17d87694d6487a95725db09aca4a49/engine.c#L104), which runs for `php.New()` and initializes the PHP VM. My starting point for this was the equivalent `embed` SAPI and the [`php_embed_init` function](https://github.com/php/php-src/blob/a7fe2570d3ce6915d4ea85c62c0f880ddc225ba7/sapi/embed/php_embed.c#L159) in particular. You may notice similarities. In a similar vein, support for binding Go method receivers as PHP classes was based on internal PHP classes, such as [curl](https://github.com/php/php-src/blob/67b4c3379a1c7f8a34522972c9cb3adf3776bc4a/ext/curl/curl_file.c).

Adding support for ZTS would involve looking at points of integration such as `engine_init` and adding equivalent `#ifdef ZTS` parts just like the PHP source does it, for example:

```c
#ifdef ZTS
  tsrm_startup(1, 1, 0, NULL);
  (void)ts_resource(0);
  ZEND_TSRMLS_CACHE_UPDATE();
#endif
```

Which should go near the top of `engine_init`, just like it appears near the top of `php_embed_init`.
Supporting ZTS across PHP 5.x and 7.x is a PITA, so I wouldn't bother doing so. Getting pre-built PHP versions with ZTS enabled is also a major PITA, so I'd suggest extending the Dockerfile to include the necessary `./configure` parameter, and use that environment for testing (via `make docker-test`).
